### PR TITLE
auth: add `oauth2` provider module

### PR DESCRIPTION
**Pull request by awanlin**
## Hey, I just made a Pull Request!

This migrates the `oauth2` provider to be a separate module

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
Hi @Rugvip, so this has pretty much everything from the current `oauth2` provider, did run into some config that wasn't in the `config.d.ts` file so I made sure to include that in this version. Also, there's one outstanding item: the current provider has a `disableRefresh` option and I'm not sure how to implement that in this module. Here's the current code: https://github.com/backstage/backstage/blob/535f5fd6ce466baf29f2c3a3c2d06b9c9c80dca3/plugins/auth-backend/src/providers/oauth2/provider.ts#L139. 

Finally, the current provider has a test that it actually triggers the auth and checks the result, should I try and replicate that?

File: .changeset/dull-eyes-pay.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-auth-backend-module-oauth2-provider': patch
```
Comment (Rugvip) [inline .changeset/dull-eyes-pay.md:2]:
```suggestion
'@backstage/plugin-auth-backend-module-oauth2-provider': minor
```

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 🎉 

Do you want to do the final step an migrating `auth-backend` to use this new module instead of the existing implementation as well?

File: plugins/auth-backend-module-oauth2-provider/src/authenticator.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      prompt: 'consent',
+    });
+  },
+
+  async authenticate(input, helper) {
+    return helper.authenticate(input);
+  },
+
+  async refresh(input, helper) {
+    return helper.refresh(input);
```
[the remark this answers is not in the document]
Comment (awanlin) [reply plugins/auth-backend-module-oauth2-provider/src/authenticator.ts:85]: Can do, just a little stuck as to how I would get at the config, any suggestions?

File: .changeset/dull-eyes-pay.md (hunk 1 above)
Comment (awanlin) [reply .changeset/dull-eyes-pay.md]: Updated 👍

**Review by Rugvip — APPROVED** [review]
Nice! 👍 Looks good, thank you! 🎉 :shipit:

**Comment by awanlin** [comment]
Thanks @Rugvip 🚀

**Pull request merged**